### DISCUSSION TITLE
Getting ltex_extra to work with LazyVim

### DISCUSSION THREAD

**Discussion by gwarf** [LazyVim/LazyVim#403] [Q&A] [answered]
Hey,
I'm aiming at doing spell check using [languagetool](https://languagetool.org/) via [ltex-ls](https://github.com/valentjn/ltex-ls) and involving [ltex_extra](https://github.com/barreiroleo/ltex_extra.nvim) to get code actions to work with LazyVim via lspconfig.

I'm trying to use the following setup (Current status with a broken workaround at https://github.com/gwarf/dotfiles/blob/main/lazyvim/nvim/lua/plugins/lsp.lua):

```lua
return {
  -- Improved ltex integration, supporting code actions
  { "barreiroleo/ltex-extra.nvim" },

  -- add various LSP to lspconfig
  {
    "neovim/nvim-lspconfig",
    -- XXX: Need to run master as current tagged release is broken and not using latest lua_ls
    -- name, cf https://github.com/neovim/nvim-lspconfig/pull/2439
    version = false,
    opts = {
      diagnostics = {
        virtual_text = false,
      },
      -- https://github.com/neovim/nvim-lspconfig/blob/master/doc/server_configurations.md
      servers = {
        -- use LanguageTool via ltex for spell checking
        -- TODO: https://dev.languagetool.org/finding-errors-using-n-gram-data.html
        ltex = {
          filetypes = {
            "bib",
            "gitcommit",
            "latex",
            "mail",
            "markdown",
            "norg",
            "org",
            "pandoc",
            "rst",
            "text",
          },
          settings = {
            -- https://valentjn.github.io/ltex/settings.html
            ltex = {
              language = "en-GB",
              additionalRules = {
                enablePickyRules = true,
                motherTongue = "fr",
              },
              -- https://community.languagetool.org/rule/list?lang=en
              disabledRules = {
                ["en-GB"] = { "TOO_LONG_SENTENCE", "OXFORD_SPELLING_Z_NOT_S", "DASH_RULE" },
                fr = { "APOS_TYP", "FRENCH_WHITESPACE", "FR_SPELLING_RULE", "COMMA_PARENTHESIS_WHITESPACE" },
              },
            },
          },
        },
      },
      setup = {
        -- FIXME: try to integrate ltex_extra with LazyVim
        -- ltex_extra fails with: Error catching ltex client
        -- the words are added to the dict, but not loaded by ltex
        ltex = function(_, opts)
          require("ltex_extra").setup({
            load_langs = { "en-GB", "fr" }, -- languages for witch dictionaries will be loaded
            init_check = true, -- whether to load dictionaries on startup
            path = vim.fn.stdpath("config") .. "/spell", -- path to store dictionaries.
            log_level = "debug", -- "none", "trace", "debug", "info", "warn", "error", "fatal"
          })
        end,
      },
    },
  },
}
```

`ltex_extra` requires to have an instance of `ltex_ls` attached to the buffer is supposed to be loaded from the `on_attach` event: https://github.com/barreiroleo/ltex_extra.nvim#configuration
I've tried to load via the setup as documented in https://github.com/LazyVim/LazyVim/blob/b1b5b461bf9d853d8472ee5b968471695118958b/lua/lazyvim/plugins/lsp/init.lua#L57-L66

The language check and the code actions works, but separately, and `ltex_extra` cannot [catch ltex](https://github.com/barreiroleo/ltex_extra.nvim/blob/aed3216f8d56d68085e2e3edd62f9236bc4d622b/lua/ltex_extra/src/commands-lsp.lua#L12-L20) and [returns an error](https://github.com/barreiroleo/ltex_extra.nvim/blob/aed3216f8d56d68085e2e3edd62f9236bc4d622b/lua/ltex_extra/src/commands-lsp.lua#L75-L76): `Error catching ltex client`.

I've seen an issue about usage with lazyvim on  ltex_extra repo, but it's apparently not configured as I'm trying to do: https://github.com/barreiroleo/ltex_extra.nvim/issues/28

Would any one have a suggestion on how this should be done in the most lazyvim standard way?

Thanks!

**Comment by r3k2** [comment]
I will also like a working solution for this, having the same issue, just migrated t o LazyVim and having issues getting this LSP to work.

**Comment by fhfuih** [comment, marked as answer]
I do have correctly configured `ltex_extra` with LazyVim (although I am not using this distro any more). There are two solutions:

-----

neovim do have a documented but less-demoed-in-examples autocommand group `LspAttach`. Using autocommands against this group is essentially the same as on_attach, but the config codes becomes more flexible. So we can do
```lua
      setup = {
        ltex = function(_, opts)
          vim.api.nvim_create_autocmd('LspAttach', function(some params)
              -- You can get server name from params. But I don't remember it. You can :h LspAttach
              -- if server name is ltex
              -- then call require("ltex_extra").setup(...)
          end)
          -- don't return true, so that the opts are used as-is by LazyVim to call lspconfig
        end,
      },
```

-----

Or, we can do

```lua
      setup = {
        ltex = function(_, opts)
          local old_on_attach = opts.on_attach
          opts.on_attach = function()
               old_on_attach()
               -- call require("ltex_extra").setup(...)
          end
          -- call lspconfig ltex setup (opts) yourself
          return true -- So that LazyVim don't call lspconfig again
        end,
      },
```

**Reply by gwarf** [reply]
Thanks a lot, it seems to work great!
I'm currently using something among this, in my `~/.config/nvim/lua/plugins/lsp.lua`:
```lua
return {
  -- Improved ltex integration, supporting code actions
  { "barreiroleo/ltex-extra.nvim" },

  -- add various LSP to lspconfig
  {
    "neovim/nvim-lspconfig",
    -- XXX: Need to run master as current tagged release is broken and not using latest lua_ls
    -- name, cf https://github.com/neovim/nvim-lspconfig/pull/2439
    version = false,
    opts = {
      -- suppress virtual text
      diagnostics = {
        virtual_text = false,
      },
      -- https://github.com/neovim/nvim-lspconfig/blob/master/doc/server_configurations.md
      servers = {
        -- use LanguageTool via ltex for spell checking
        ltex = {
          filetypes = {
            "bib",
            "gitcommit",
            "latex",
            "mail",
            "markdown",
            "norg",
            "org",
            "pandoc",
            "rst",
            "text",
          },
          settings = {
            -- https://valentjn.github.io/ltex/settings.html
            ltex = {
              language = "en-GB",
              additionalRules = {
                enablePickyRules = true,
                motherTongue = "fr",
              },
              -- https://community.languagetool.org/rule/list?lang=en
              disabledRules = {
                ["en-GB"] = { "TOO_LONG_SENTENCE", "OXFORD_SPELLING_Z_NOT_S", "DASH_RULE" },
                fr = { "APOS_TYP", "FRENCH_WHITESPACE", "FR_SPELLING_RULE", "COMMA_PARENTHESIS_WHITESPACE" },
              },
            },
          },
        },
      },
      setup = {
        -- integrate ltex_extra with lazyvim
        -- https://github.com/LazyVim/LazyVim/discussions/403
        ltex = function(_, opts)
          vim.api.nvim_create_autocmd("LspAttach", {
            callback = function(args)
              local client = vim.lsp.get_client_by_id(args.data.client_id)
              if client.name == "ltex" then
                require("ltex_extra").setup({
                  load_langs = { "en-GB", "fr" }, -- languages for witch dictionaries will be loaded
                  init_check = true, -- whether to load dictionaries on startup
                  path = vim.fn.stdpath("config") .. "/spell", -- path to store dictionaries.
                  log_level = "none", -- "none", "trace", "debug", "info", "warn", "error", "fatal"
                })
              end
            end,
          })
        end,
      },
    },
  },
}
```

I can use the quick action to add words to the dictionary or fix them, and they are correctly loaded on subsequent runs, so seems to address my current needs perfectly.
Thanks again!

**Comment by weakish** [comment]
https://github.com/barreiroleo/ltex_extra.nvim/issues/28 has been closed. Right now using ltex_extra is very straight forward:

```lua
-- ~/.config/nvim/lua/plugins/ltex.lua

return {
  {
    "barreiroleo/ltex_extra.nvim",
    dependencies = { "neovim/nvim-lspconfig" },
  },
}
```

**Reply by gwarf** [reply]
Have you been able to get this kind of declaration while passing some ltex and ltex_extra options in a LazyVim based setup?

**Reply by weakish** [reply]
I did not tested it before. And recently I find out that this simple configuration stopped to work. Not sure it is due to Neovim/LazyVim/ltex-extra updates or conflict with my other configurations of LSP. I changed to the LspAttach way ([configuration posted by gwarf above](https://github.com/LazyVim/LazyVim/discussions/403#discussioncomment-5319825)) instead.

**Comment by dimas-cyriaco** [comment]
@gwarf @fhfuih This config is still working for you? I'm trying it, but the code actions for adding to the dictionary does not work.

**Reply by gwarf** [reply]
I'm currently using this configuration that works quite OK for me: https://github.com/gwarf/dotfiles/blob/20bbad28426398f98b97764bc87100a2f6b72d32/lazyvim/nvim/lua/plugins/lsp.lua

**Reply by gwarf** [reply]
I've some usability issues, like having checks being triggered a bit too often or slowing things a bit too much, but I use it daily for markdown, neorg and emails files.
I also noticed that if I open the Lazy window too quickly when loading a big neorg file, it will prevent the proper configuration to load correctly, and I will get the default one, with some errors like:
```
Error executing vim.schedule lua callback: ...vim/lazy/ltex_extra.nvim/lua/ltex_extra/commands-lsp.lua:84: Error catching ltex client
stack traceback:
	[C]: in function 'updateConfig'
	...vim/lazy/ltex_extra.nvim/lua/ltex_extra/commands-lsp.lua:84: in function <...vim/lazy/ltex_extra.nvim/lua/ltex_extra/commands-lsp.lua:83>
Error executing vim.schedule lua callback: ...vim/lazy/ltex_extra.nvim/lua/ltex_extra/commands-lsp.lua:84: Error catching ltex client
stack traceback:
	[C]: in function 'updateConfig'
	...vim/lazy/ltex_extra.nvim/lua/ltex_extra/commands-lsp.lua:84: in function <...vim/lazy/ltex_extra.nvim/lua/ltex_extra/commands-lsp.lua:83>
```
But as I'm aware of this I'm just trying to avoid starting `:Lazy` too quickly (via `<Leader>l`), and it's OK.

**Reply by gwarf** [reply]
In fact, I think that the command run is not relevant (`:Laay` here), it's just that a command spanning a window/dialog is then confusing ltex_extra.

**Reply by barreiroleo** [reply]
Hello there! First of all, sorry for the issues and thanks for supporting each other.

I'm planning to merge a new release after Nvim 0.10 release and (hopefully) all these kind of problems should be resolved (or at least more easy to resolve).
It's available on the dev branch if you want to try it. The setup was simplified a lot and made more standard.
Also, there are a couple of options marked as deprecated, but I still supporting them as long the changes lands successfully. For these options I also added notes about how to migrate in the Readme.

Many thanks for using this tool and if you have any concern, please raise an issue. (I may be slow to respond but I will).
https://github.com/barreiroleo/ltex_extra.nvim/tree/dev
